### ISSUE TITLE
firestore: Switch to the rbtree3 gem

### ISSUE THREAD

**Issue by knu** [googleapis/google-cloud-ruby#8421]
**Is your feature request related to a problem? Please describe.**
The original rbtree library has not been updated since 2013.  It has compatibility problems with recent versions of Ruby.  It produces deprecation warnings on Ruby 2.7 and is not friendly with MJIT and Compaction GC, and it does not even build with Ruby 3.0 soon to be released.

 **Describe the solution you'd like**
To fix them, I suggest switching to the fork called [rbtree3](https://github.com/kyrylo/rbtree3).  It is used by the [ruby client](https://github.com/airbrake/airbrake-ruby) of [Airbrake](https://airbrake.io/) and developed by the vendor.

I confirmed it builds and works with the latest snapshot of Ruby 3.0.0dev.

https://github.com/knu/google-cloud-ruby/commit/use_rbtree3

 **Describe alternatives you've considered**

 **Additional context**

**Comment by mlarraz** [comment]
Seconding this.

The Ruby stdlib's `Set` class will attempt to require `rbtree` if it's available: https://github.com/ruby/ruby/blob/d0693391103672b7fc3976e5ae8ae414059b3e97/lib/set.rb#L699-L704

This means any project running Ruby 2.7 will produce countless warnings of
```
warning: rb_safe_level will be removed in Ruby 3.0
```

On one of our applications this volume was significant enough to impact request latency.


A [new version](rubygems.org/gems/rbtree) of the gem was recently released which addresses this. ([Full Diff](my.diffend.io/gems/rbtree/0.4.2/0.4.4))

However any projects depending on this gem clearly need to move off it, as I cannot even find a working public repository for it.

**Comment by quartzmo** [comment]
The original `rbtree` is [showing signs of life](https://www.rubydoc.info/gems/rbtree/0.4.2#label-0.4.1), with two releases this month. Does anyone know if this signals a roadmap to Ruby 3.0 support?

**Comment by quartzmo** [comment]
@dazuma Do you approve changing the dependency to [rbtree3](https://github.com/kyrylo/rbtree3)?

**Comment by dazuma** [comment]
I don't have a strong opinion on this. I looked at rbtree and rbtree3 and am not terribly happy with the history/activity on either. But I trust @knu 's judgment on this and would be okay switching to rbtree3.

**Comment by quartzmo** [comment]
I'm having problems using `rbtree3` as a drop-in replacement for `rbtree`, I think because of [this commit](https://github.com/kyrylo/rbtree3/commit/f76a7202d41a81bb9e20f9b9e16cabd4fea10605), notably:

> With this change we no longer support smooth transition between RBTree and Hash,
which preserves default value. I'm not 100% sure that dropping support is the
right way but there's only much I can do. My Ruby C API knowledge is not very
deep.

> This can potentially break someone's code but it doesn't seem to be too
critical.

The issues I've hit so far are:

* `google-cloud-ruby/google-cloud-firestore/lib/google/cloud/firestore/watch/inventory.rb:268:in `remove_doc_from_tree': undefined method `key' for #<RBTree:0x00007fb9559e0118> (NoMethodError)`
* `google-cloud-ruby/google-cloud-firestore/lib/google/cloud/firestore/watch/inventory.rb:162:in `query_comparison_proc': undefined method `ref' for "projects/projectID/databases/(default)/documents/watch/doc":String (NoMethodError)`

**Comment by quartzmo** [comment]
Using `rbtree` `0.4.4` with ruby `2.7.2p137`, I do not see any deprecation warnings while running `rake ci` and `rake acceptance` for google-cloud-firestore.

**Comment by knu** [comment]
Thanks for the feedback, and sorry for not responding sooner!

> I'm having problems using `rbtree3` as a drop-in replacement for `rbtree`, I think because of [this commit](https://github.com/kyrylo/rbtree3/commit/f76a7202d41a81bb9e20f9b9e16cabd4fea10605), notably:
> 
> > With this change we no longer support smooth transition between RBTree and Hash,
> > which preserves default value. I'm not 100% sure that dropping support is the
> > right way but there's only much I can do. My Ruby C API knowledge is not very
> > deep.
> 
> > This can potentially break someone's code but it doesn't seem to be too
> > critical.

I believe it got fixed in https://github.com/kyrylo/rbtree3/commit/38e7277f26e75c190c3adb2d4f2b3786b0fc0d25.

> The issues I've hit so far are:
> 
> * `google-cloud-ruby/google-cloud-firestore/lib/google/cloud/firestore/watch/inventory.rb:268:in `remove_doc_from_tree': undefined method `key' for #<RBTree:0x00007fb9559e0118> (NoMethodError)`
> * `google-cloud-ruby/google-cloud-firestore/lib/google/cloud/firestore/watch/inventory.rb:162:in `query_comparison_proc': undefined method `ref' for "projects/projectID/databases/(default)/documents/watch/doc":String (NoMethodError)`

Do they work with the original rbtree?  The first error comes from the line calling `key` on a RBTree object, but there should be no RBTree#key.  The second error seems to be caused by RBTree#readjust calling the comparator passing strings where DocumentSnapshot objects (keys of the RBTree object) are expected.  Would you provide a bit more context?

**Comment by quartzmo** [comment]
@knu This is what I observe regarding `RBTree#key`.

### rbtree-0.4.4
```
quartzmo@Chriss-iMac google-cloud-firestore % ruby -v
ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-darwin19]
quartzmo@Chriss-iMac google-cloud-firestore % bundle exec gem which rbtree
/Users/quartzmo/.rbenv/versions/2.7.2/gemsets/quartzmo/gems/rbtree-0.4.4/lib/rbtree.bundle
quartzmo@Chriss-iMac google-cloud-firestore % bundle exec irb
irb(main):001:0> require "rbtree"
=> true
irb(main):002:0> rbtree = RBTree["c", 10, "a", 20]
=> #<RBTree: {"a"=>20, "c"=>10}, default=nil, cmp_proc=nil>
irb(main):003:0> rbtree.key 10
=> "c"
```

### rbtree3-0.6.0
```
quartzmo@Chriss-iMac google-cloud-firestore % bundle exec gem which rbtree
/Users/quartzmo/.rbenv/versions/2.7.2/gemsets/quartzmo/gems/rbtree3-0.6.0/./rbtree.bundle
quartzmo@Chriss-iMac google-cloud-firestore % bundle exec irb
irb(main):001:0> require "rbtree"
=> true
irb(main):002:0> rbtree = RBTree["c", 10, "a", 20]
=> #<RBTree: {"a"=>20, "c"=>10}, default=nil, cmp_proc=nil>
irb(main):003:0> rbtree.key 10
Traceback (most recent call last):
        4: from /Users/quartzmo/.rbenv/versions/2.7.2/bin/irb:23:in `<main>'
        3: from /Users/quartzmo/.rbenv/versions/2.7.2/bin/irb:23:in `load'
        2: from /Users/quartzmo/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/irb-1.2.6/exe/irb:11:in `<top (required)>'
        1: from (irb):3
NoMethodError (undefined method `key' for #<RBTree: {"a"=>20, "c"=>10}, default=nil, cmp_proc=nil>)
Did you mean?  keys
               key?
```

**Comment by knu** [comment]
Thanks, I failed to confirm that correctly.  Today I checked out and compared the two gems, to find out rbtree3 is a fork from rbtree 0.3.0.  The original rbtree, later in 0.4.x series, added some methods such as RBTree#key, RBTree#default_proc= and RBTree#select! to align with Ruby's builtin Hash API, but rbtree3 did not incorporate them.  This is why rbtree3 is missing RBTree#key.

So, while rbtree3 is more active in development and eager to quickly follow changes in Ruby, rbtree has at least caught up.  I think it's a safer option for this project to just update rbtree to 0.4.4 for now.
